### ISSUE TITLE
Vary instructions by lane based on destination:lanes

### ISSUE THREAD

**Issue by 1ec5** [Project-OSRM/osrm-backend#4222]
A route step’s `instruction` and `destinations` should account for the `destination:lanes` and `destination:ref:lanes` tags when present. Consider the following scenario:

[![dearborn2](https://user-images.githubusercontent.com/1231218/27760810-277b9be8-5e05-11e7-8f14-6568f83b2e5c.jpg)](https://www.mapillary.com/map/im/mUDqicHyq4in0_OMpYYdkw)

<img width="840" alt="iD" src="https://user-images.githubusercontent.com/1231218/27760241-724e6ed8-5df7-11e7-974d-6b96aa70360f.png">

with this [collector-distributor way](http://www.openstreetmap.org/way/431431064):

```
destination=Dearborn Street;James Street;Madison Street;Spokane
destination:lanes=Dearborn Street;James Street;Madison Street|Spokane
destination:ref=I 90 East
destination:ref:lanes=|I 90 East
```

If the route travels to Dearborn Street, then the route step representing the collector-distributor way should have a `destinations` of `Dearborn Street;James Street;Madison Street` and an `instruction` mentioning Dearborn Street. On the other hand, if the route travels towards Spokane on eastbound I-90, then that same route step should have a `destinations` of `I 90 East: Spokane` and an `instruction` that mentions Spokane.

https://github.com/Project-OSRM/osrm-backend/issues/2553#issuecomment-227067935 suggests that `destination:lanes` could help deduce the correct lane to take.

/ref mapbox/mapping#308 Project-OSRM/osrm-text-instructions#30
/cc @freenerd @daniel-j-h

**Comment by 1ec5** [comment]
In the United States, collector-distributor lanes are common in urban areas and at interchanges between freeways. These are often critical junctions where missing a turn could mean a 10-minute detour. OSRM’s current behavior of observing `destination` and `destination:ref` but not `destination:lanes` or `destination:ref:lanes` can cause clients such as the Mapbox Navigation SDK for iOS to display mismatched UI.

For example, here’s an exit in which the left two lanes (out of five) share a single exit and split shortly thereafter (from [Mapillary](https://www.mapillary.com/map/im/HJYyzBAFTmSfS6F_5tBgOg)):

[![1b-1a](https://user-images.githubusercontent.com/1231218/27992738-cc956136-644f-11e7-9bcb-3aa8d15c82f0.jpg)](https://www.mapillary.com/map/im/HJYyzBAFTmSfS6F_5tBgOg)

[The way](http://www.openstreetmap.org/way/19051476) shared between exits 1B and 1A is tagged:

```
destination=2nd Street;Downtown;Riverfront
destination:ref=I 71 North;US 50 East
destination:lanes=|2nd Street;Downtown;Riverfront
destination:ref:lanes=I 71 North;US 50 East|
destination:to=I 471;US 52 East
destination:ref:to:lanes=I 471;US 52 East|
junction:ref=1B;1A
junction:ref:lanes=1B|1A
```

[This route](http://map.project-osrm.org/?z=18&center=39.102959%2C-84.524158&loc=39.103694%2C-84.526727&loc=39.101800%2C-84.523154&hl=en&alt=0) would first take the shared 1B-1A exit, then exit 1A towards 2nd Street (as opposed to I-71). OSRM returns an `exits` of `1B;1A` and a `destination` of `I 71 North, US 50 East: 2nd Street, Downtown, Riverfront`, which looks like this in a typical client:

<img src="https://user-images.githubusercontent.com/1231218/27992741-e358d8f8-644f-11e7-911e-3c47de11508b.png" alt="mapbox-navigation-ios" width="300">

This is a problem, because the label suggests one subexit and the shield suggests another. The expected behavior would be a `destination` of just `2nd Street, Downtown, Riverfront`, based on the right side of `destination:lanes` and the right side of `destination:ref:lanes` (the empty string).

The way in question probably should be tagged with `lanes=2`, but I wonder if the number of lanes could be inferred from `destination:lanes`.

/cc @willwhite @ericrwolfe

**Comment by 1ec5** [comment]
In this same interchange, [a route leading to 2nd Street](http://map.project-osrm.org/?z=16&center=39.102332%2C-84.523036&loc=39.103694%2C-84.526727&loc=39.099510%2C-84.521851&hl=en&alt=0) (exit 1A) ends up saying “take exit 1B” and never says “take exit 1A”. This is due to a combination of factors:

* [The `highway=motorway_link` way shared](http://www.openstreetmap.org/way/19051476) between the two exits is tagged `junction:ref=1B;1A`, but OSRM Text Instructions discards all but the first exit number, saying only `1B`.
* [The subsequent `highway=motorway_junction` node](http://www.openstreetmap.org/node/339939162) where exits 1B and 1A split is treated as a fork in the road by OSRM (“Keep right at the fork”), rather than a highway exit, even though the individual `highway=motorway_link` ways after that are tagged with the correct `junction:ref` tags.

**Comment by 1ec5** [comment]
Perhaps OSRM Text Instructions could be responsible for varying the instructions based on `destination:lanes` and `destination:ref:lanes`. The backend would need to expose those tags’ values in the Intersection objects; beyond that, it should be straightforward for OSRM Text Instructions to match a lane destination with a lane indication, assuming there are lane indications in the first place.

**Comment by MoKob** [comment]
@1ec5 when exposing `destination:lanes`, I would kind of expect a connection to `turn:lanes` as provided in the `lanes` field of our response. Do you imagine such a relation, or would it be Ok to just return these `destination:lanes` as we do in destination?
In case we need to to `destination:lanes` as `lanes`, exposing these would become way more complicated, since we would need name IDs for on lane basis, for which I cannot even guess how much work that would be right now.

**Comment by 1ec5** [comment]
I think OSRM Text Instructions could make use of a raw `destination:lanes` value that includes all the lanes, provided that lane indications are available for that same intersection. Suppose a way has the following tags:

```
destination:lanes=SR 4 North|SR 4 North;SR 4 South
turn:lanes=through|through;right
```

If both lanes are valid, as indicated in the Intersection object, then OSRM Text Instructions could pick `SR 4 North` from `destination:lanes` to use in place of `destination`. If the route uses the right lane only, then it could pick out `SR 4 South` the same way.

**Comment by willwhite** [comment]
Here's a case of a collector-distributor where adding lanes wouldn't help:

- [Should be toward San Francisco - instead says San Jose](https://www.mapbox.com/get-directions/#13.96/37.3223/-122.0497?coordinates=-122.049736,37.322295;-122.066717,37.334422)
- [Correctly says toward San Jose](https://www.mapbox.com/get-directions/#13.96/37.3223/-122.0497?coordinates=-122.049736,37.322295;-122.0430770799443,37.334094498715274)

I'm not sure this was proposed as such, but this wouldn't be an alternative for https://github.com/Project-OSRM/osrm-text-instructions/issues/30.

**Comment by 1ec5** [comment]
> I'm not sure this was proposed as such, but this wouldn't be an alternative for Project-OSRM/osrm-text-instructions#30.

I agree, Project-OSRM/osrm-text-instructions#30 would be desirable for precisely the situation you describe. This issue tracks a related but distinct feature, in which we not only include the most relevant ref and destination but also vary the suggested lanes, which will affect the instruction as well (“Use the second lane from the right to take the ramp…”).

**Comment by 1ec5** [comment]
As an alternative to implementing the proposed logic in OSRM-backend, the `destination:lanes`, `destination:street:lanes`, and `destination:ref:lanes` values could be added to Intersection objects alongside the existing `turn:lanes` values. A client of OSRM could then implement the disambiguation logic itself.

/ref https://github.com/Project-OSRM/osrm-backend/issues/3281#issuecomment-338807712
